modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -691,40 +687,103 @@ public String getKey() {
         Operation(String key) {
             this.key = key;
         }
+
+        public static Operation fromKey(String key) {
+            for (Operation operation : Operation.values()) {
+                if (operation.key.equals(key)) {
+                    return operation;
+                }
+            }
+            throw new IllegalArgumentException("No matching key: " + key);
+        }
     }
 
-    private record StatsKey(Operation operation, OperationPurpose purpose) {
+    // visible for testing
+    record StatsKey(Operation operation, OperationPurpose purpose) {
         @Override
         public String toString() {
             return purpose.getKey() + "_" + operation.getKey();
         }
     }
 
-    private static class StatsCollectors {
-        final Map<StatsKey, LongAdder> collectors = new ConcurrentHashMap<>();
+    // visible for testing
+    class StatsCollectors {
+        final Map<StatsKey, OperationMetrics> collectors = new ConcurrentHashMap<>();
 
         Map<String, Long> statsMap(boolean stateless) {
             if (stateless) {
                 return collectors.entrySet()
                     .stream()
-                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().sum()));
+                    .collect(Collectors.toUnmodifiableMap(e -> e.getKey().toString(), e -> e.getValue().count()));
             } else {
                 Map<String, Long> normalisedStats = Arrays.stream(Operation.values()).collect(Collectors.toMap(Operation::getKey, o -> 0L));
                 collectors.forEach(
                     (key, value) -> normalisedStats.compute(
                         key.operation.getKey(),
-                        (k, current) -> Objects.requireNonNull(current) + value.sum()
+                        (k, current) -> Objects.requireNonNull(current) + value.count()
                     )
                 );
                 return Map.copyOf(normalisedStats);
             }
         }
 
-        public void onSuccessfulRequest(Operation operation, OperationPurpose purpose) {
-            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new LongAdder()).increment();
+        public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
+            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new OperationMetrics(operation, purpose))
+                .onRequestComplete(requestMetrics);
         }
     }
 
+    // visible for testing
+    class OperationMetrics {
```

Comment:
Nit: can we call this something like `OperationMetricsRecorder` (or `StatsCollector` singular) since we also have `RequestMetrics` which is a data class and this one is not?

Reply:
Updated this in a commit where I addressed @mhl-b's feedback also: f04831e3b071b921e92e359674554d4b80bfa989